src/masking/mod.rs

```diff
@@ -1 +1,91 @@
+use core::mem::MaybeUninit;
+use core::num::{NonZeroUsize, Wrapping};
+
+use crate::capacity::MaskingCapacity;
+use crate::storage::{PartialStorage, Storage};
+
 mod tests;
+
+pub struct MaskingRingBuffer<S: PartialStorage<Capacity = MaskingCapacity>> {
+    /// The underlying storage
+    storage: MaybeUninit<S>,
+    /// The start of the buffer in the storage (`0..CAPACITY`)
+    // TODO: Benchmark whether Wrapping is worth it
+    index: Wrapping<usize>,
+    /// The number of items in the buffer (`0..=CAPACITY`)
+    len: usize,
+}
+
+impl<S: PartialStorage<Capacity = MaskingCapacity>> MaskingRingBuffer<S> {
+    pub fn new(storage: S) -> Self {
+        MaskingRingBuffer {
+            storage: MaybeUninit::new(storage),
+            index: Wrapping(0),
+            len: 0,
+        }
+    }
+
+    /// Returns whether the ringbuffer is full
+    ///
+    /// A ringbuffer is full when its length equals its capacity. If an item is enqueued
+    /// while the ringbuffer is full [MaskingRingBuffer::enqueue] will dequeue an item to
+    /// make room for the new item. The dequeued item will be returned.
+    pub fn is_full(&self) -> bool {
+        self.len == self.capacity().get()
+    }
+
+    /// Returns true when the ringbuffer is empty
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// The capacity of the underlying storage
+    ///
+    /// This is the maximum number of items that the ringbuffer can hold.
+    pub fn capacity(&self) -> NonZeroUsize {
+        unsafe { Storage::capacity(self.storage.assume_init_ref() as *const S) }.into()
```

Comment:
I think `MaybeUninit::as_ptr()` will be nicer to use here.

Reply:
Cool, will change all of them (and the others to as_mut)